From b902a1b789b52931994f785d83998b81df3ec1ab Mon Sep 17 00:00:00 2001
From: Andrew Eikum <aeikum@codeweavers.com>
Date: Thu, 29 Mar 2018 14:14:23 -0500
Subject: [PATCH] xaudio2: Prefer native xaudio2 in 32-bit

---
 dlls/xaudio2_7/xaudio_dll.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/xaudio2_7/xaudio_dll.c b/dlls/xaudio2_7/xaudio_dll.c
index 28cf98565b..08c1ec1e84 100644
--- a/dlls/xaudio2_7/xaudio_dll.c
+++ b/dlls/xaudio2_7/xaudio_dll.c
@@ -89,6 +89,10 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD reason, void *pReserved)
 
     switch (reason)
     {
+#ifndef _WIN64
+    case DLL_WINE_PREATTACH:
+        return FALSE;  /* prefer native version */
+#endif
     case DLL_PROCESS_ATTACH:
         instance = hinstDLL;
         DisableThreadLibraryCalls( hinstDLL );
